### DISCUSSION TITLE
GetMolHash in Java

### DISCUSSION THREAD

**Discussion by youngmip** [rdkit/rdkit#6660] [Q&A]
Hi, 

I managed to create Java wrapper suing SWIG and added the Jar and native library to my project. 
Now, I need to write this Python code using the Java wrapper. 
```
 from rdkit import Chem
 from rdkit.Chem import RegistrationHash
 from rdkit.Chem.RegistrationHash import HashLayer
...
    m = Chem.MolFromMolBlock(ctab, removeHs=False, sanitize=True)

    h = RegistrationHash.GetMolLayers(m)

    tautomer_insensitive_hash = RegistrationHash.GetMolHash(h, RegistrationHash.HashScheme.TAUTOMER_INSENSITIVE_LAYERS)

    all_layer_hash = RegistrationHash.GetMolHash(h)

    stereo_insensitive_hash = RegistrationHash.GetMolHash(h,RegistrationHash.HashScheme.STEREO_INSENSITIVE_LAYERS)
```

However, I can't find appropriate Java API to use. 
The generated jar and native library were found under $RDKIT/Code/JavaWrappers/gmwrapper. 

Can you please help for this? 

Thanks in advance!

Regards,
Youngmi

**Comment by sheeehy** [comment]
You need to import the classes from your Java wrapper and make sure the JAR and native library are added to your Java project, 

```
import org.RDKit.*;

public class RDKitExample {
    public static void main(String[] args) {
  
        // Read a molecule from an MDL MolBlock,
        String ctab = "your_molblock_here";
        ROMol m = RWMol.MolFromMolBlock(ctab, false, true);

        // Get the molecular layers
        RDKFuncs.RDKFuncsInit();
        ROMol h = RegistrationHash.GetMolLayers(m);

        // tautomer-insensitive hash
        String tautomerInsensitiveHash = RegistrationHash.GetMolHash(h, RegistrationHash.HashScheme.TAUTOMER_INSENSITIVE_LAYERS);

        // all layer hash
        String allLayerHash = RegistrationHash.GetMolHash(h);

        // stereo-insensitive hash
        String stereoInsensitiveHash = RegistrationHash.GetMolHash(h, RegistrationHash.HashScheme.STEREO_INSENSITIVE_LAYERS);

        // You can use the hash values as needed
        System.out.println("Tautomer-Insensitive Hash: " + tautomerInsensitiveHash);
        System.out.println("All Layer Hash: " + allLayerHash);
        System.out.println("Stereo-Insensitive Hash: " + stereoInsensitiveHash);
    }
}
```

then replace "your_molblock_here" with your actual MDL MolBlock string.

**Reply by sitanshubhunia** [reply]
@sheeehy 

Can you provide the required jar files to run this java. As whenever I added the jar(s), getting that there is no class RegistrationHash . I am try to run this code from windows environment. Also I have added required dll file

**Reply by youngmip** [reply]
@sheeehy Thank you for the help! 

When I tested your code, I found the same issue @sitanshubhunia  found. As you see RegistrationHash class doesn't exist and  RDKFuncs.RDKFuncsInit() is unavailable.
![image](https://github.com/rdkit/rdkit/assets/11573733/4a9e9db9-2886-43d0-bba2-86dbef7ef4dd)

I created a jar and a native library from the following release:
https://github.com/rdkit/rdkit/releases/tag/Release_2023_03_3

**Comment by ptosco** [comment]
Small addition to the above: remember to call the `delete()` method on C++ objects when you do not need them anymore (i.e., `m.delete()`, `h.delete()`), since they are _not_ going to be garbage collected like normal Java objects.

**Comment by ptosco** [comment]
You may find relatively recent `jar` files here:
https://github.com/rdkit/rdkit-build-helpers/tree/main/swig_wrappers/artifacts/java

**Reply by youngmip** [reply]
Hi @ptosco 
thank you for the additional help. 
However, I still have a problem that I can't find the RegistrationHash class and the RDKFuncs.RDKFuncsInit() method from both Jar files that I created from the release 2023_03_3 and that I found from this link https://github.com/rdkit/rdkit-build-helpers/tree/main/swig_wrappers/artifacts/java/noarch.

Can you please help? Many thanks in advance.
